Refacto layer

peppsac opened the pull request:
(Last) core refactoring of itowns.

Most important commit is: c0cfb0b7a74fb6ebe6af9b7e0cde8704f2fdde95
```
    refactor: remove Scene and reorganize files
    
    Split Scene responsabilities between:
      * View: container for datas
      * MainLoop: itowns' main loop (update/render)
    
    GlobeView is a specialized View, ready for displaying a Globe.
    c3DEngine also loses Globe-related code, which moves to GlobeView.
    
    Files/folders hierarchy has changed as well.
```

This PR also adds 2 examples:
* `planar.html`: itowns in planar mode (EPSG:3946)
* `multiglobe.html`: shows 2 itowns globes on the same page (note: debug and controls are buggy in this example)

peppsac committed:
refactor: move globe specific code to ApiGlobe

peppsac committed:
feature: add DEFINITIVE_ERROR state to LayerUpdateState

Prevents retrying when we know that a given request will never succeed.

peppsac committed:
refactor: remove Scene and reorganize files

Split Scene responsabilities between:
  * View: container for datas
  * MainLoop: itowns' main loop (update/render)

GlobeView is a specialized View, ready for displaying a Globe.
c3DEngine also loses Globe-related code, which moves to GlobeView.

Files/folders hierarchy has changed as well.

peppsac committed:
feature: add a new prefab: PlanarView

This change requires to expose more of itowns to html.
I only exposed what I needed to build the example page.

peppsac committed:
refactor: remove WMTS dependencies in Layer update code

Use BoundingBox to represent a SIG-surface and store this in textures.
This way we can compute offset/scale for every supported coordinate set.

peppsac committed:
refactor: rename BoundingBox -> Extent

This also changes the semantic of this object: it's now a SIG surface

peppsac committed:
feature: add support for color textures as elevation

peppsac committed:
feature: add option to disable tilemesh skirt

peppsac committed:
fix: don't try to compute min/max elevation if no bufffer

peppsac committed:
fix: cancel subdivision commands if parent is deleted

peppsac committed:
fix: add tolerance when comparing Extent

peppsac committed:
example: add MultiGlobe example

Jeremy-Gaillard commented on src/Core/Prefab/PlanarView.js:
```diff
@@ -0,0 +1,262 @@
+PlanarView.prototype = Object.create(View.prototype);
+PlanarView.prototype.constructor = PlanarView;
+
+PlanarView.prototype.addLayer = function addLayer(layer) {
```
I have an issue with the overriding of this function. If I want to add a layer that is independent from the plane itself (a 3d-tile layer in my case), I no longer can add it to the view without attaching it to the terrain tile layer.

peppsac replied:
Yes you can, using:
```js
itowns.View.prototype.addLayer.call(view, pointcloud);
```
The syntax is special, but it's an advanced use case so I don't mind it.

Jeremy-Gaillard commented on src/Main.js:
```diff
@@ -1,19 +1,26 @@
- * To change this template file, choose Tools | Templates
- * and open the template in the editor.
- */
+// Modules using `default` export must be imported...
```
Some exports are missing (GeometricLayer for example). I think a good test to see if enough classes are exposed is to try to recreate PlanarView or GlobeView from outside the core.

peppsac replied:
I'll add an advanced example without using `PlanarView` or `GlobeView` so I'll need to update exports.
It might end up in a separate PR though.

Jeremy-Gaillard commented on src/Process/GlobeTileProcessing.js:
```diff
@@ -1,9 +1,9 @@
+import Extent from '../Core/Geographic/Extent';
 
 const cV = new THREE.Vector3();
 let vhMagnitudeSquared;
```
Won't these global variables mess things up when having multiple instances of iTowns? It shouldn't affect anything if the instances have the same window size though.

peppsac replied:
Yes this is a problem.
This is the same reason that makes the globe controls behave weirdly in the `multiglobe` example.

Jeremy-Gaillard reviewed (commented):
Just a few things I noticed while working on the 3d-tiles branch, I haven't actually read the whole PR.

peppsac committed:
feature: add isBox3DVisible method to Camera

Instead of implementing visilibty testing twice in Planar and Globe.

This commit also improve the test, by cancelling the translation component
of camera's matrix. This way we avoid numerical errors due to using big numbers.

gchoqueux commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -1,11 +1,75 @@
+
+        const { min, max } = useMinMaxFromParent ?
+        {
+            min: parent.OBB().z.min,
```
Seems useless because already done [here](https://github.com/iTowns/itowns2/pull/301/files#diff-48576d3351459435df139556b817119dR65)

peppsac replied:
Thanks, fixed.

gchoqueux commented:
In Viewer, how remove layer and ordered layer?

gchoqueux commented:
Bugs :
* Imagery layer are downscaled (some time)
* problem with right clic (context menu appears)
* The selection of tiles doesn't work all the time (May reveal a refresh problem)

gchoqueux commented:
don't forget to update `examples/index.hmtl`

gchoqueux commented:
It seems to have problems with `colorLayersId` and`layerTexturesCount` 

`colorLayersId` can have unnecessary  color layers -> The `LayeredMaterialNodeProcessing` must have introduced an error.

Example :
For one tile on france with tile's zoom === 3
   * tile.material.colorLayersId === ["Ortho", "OrthoCRS", "ScanEX", "Region"]
   * tile.material.layerTexturesCount === [3, 1, 3, 1]

While the layer `OrthoCRS` starts at zoom 6

peppsac commented:
New update pushed.
The issues that should be fixed are:
* Imagery layer are downscaled (from https://github.com/iTowns/itowns2/pull/301#issuecomment-301469083)
* The selection of tiles doesn't work all the time (May reveal a refresh problem)  (from https://github.com/iTowns/itowns2/pull/301#issuecomment-301469083)
* Some exports are missing (GeometricLayer for example). (from https://github.com/iTowns/itowns2/pull/301#discussion_r116273196)
* don't forget to update examples/index.hmtl (from https://github.com/iTowns/itowns2/pull/301#issuecomment-301471619)

peppsac commented:
Pushed a fix for: 

- colorLayersId can have unnecessary color layers (from https://github.com/iTowns/itowns2/pull/301#issuecomment-301475771)

gchoqueux commented:
**bug** resize windows : the layers aren't updated before rendering (the frustrum culling is updated after rendering)

peppsac committed:
fix: Fix LayeredMaterial.setSequence method

moveElementsArray wasn't working reliably since it depends on offset/position,
which are being changed at the same time.
So use a simpler method: copy the array first, and copy using oldOffset from
this copy.

This commit also add a Debug tool to check texture / layers consistency.

peppsac commented:
New update. I've fixed a few bugs regarding layers ordering.

@gchoqueux regarding "In Viewer, how remove layer and ordered layer?" : see how's ApiGlobe is doing it now.

peppsac committed:
feature: simplify redraw handling

Now that the underlying structure has been cleaned up, we can simplify
redraw handling.
Redraws are performed if requested, and only if requested.
We also stop looping (ie: step() is never called from step()).

peppsac commented:
New update and "bug resize windows : the layers aren't updated before rendering (the frustrum culling is updated after rendering)" (from https://github.com/iTowns/itowns2/pull/301#issuecomment-301512331) is fixed.

gchoqueux commented:
Tile's DMT can flick when user zoom and stop navigation

gchoqueux commented:
Why realistic lighting is broken?

gchoqueux commented:
2 problems with GUI:
* Right clic can make appear a context menu (instead of paning)
* If we close, open Gui controls and move globe : can make select text in Gui (instead of moving globe)

peppsac commented:
Update++
Issues fixed:
* Tile's DMT can flick when user zoom and stop navigation (https://github.com/iTowns/itowns2/pull/301#issuecomment-301766087)
* Why realistic lighting is broken? (https://github.com/iTowns/itowns2/pull/301#issuecomment-301766223)

I can't reproduce the 2 "GUI" issues  from https://github.com/iTowns/itowns2/pull/301#issuecomment-301767534 - can you still reproduce them?

gchoqueux commented:
All Layer.sequence aren't updated when we remove imagery layer

peppsac commented:
"All Layer.sequence aren't updated when we remove imagery layer" (https://github.com/iTowns/itowns2/pull/301#issuecomment-301810085) fixed as well

gchoqueux commented on src/Renderer/c3DEngine.js:
```diff
@@ -7,205 +7,73 @@
-
-        this.renderer.clear();
+    this.renderScene = function renderScene(camera) {
+        if (camera.camHelper())
```
remove camHelper()

peppsac replied:
done

gchoqueux commented on src/Renderer/c3DEngine.js:
```diff
@@ -7,205 +7,73 @@
     }.bind(this);
 
     this.onWindowResize = function onWindowResize() {
         this.width = this.viewerDiv.clientWidth * (this.debug ? 0.5 : 1);
```
replace by `this.width = this.viewerDiv.clientWidth`

peppsac replied:
done

gchoqueux commented on src/Core/View.js:
```diff
@@ -0,0 +1,185 @@
+    this._layers = [];
+
+    if (__DEBUG__) {
+        this._debug = new Debug(this, viewerDiv);
```
this._debug is used?

peppsac replied:
Nope and Debug is a plain function, not a constructor so  I need to this this... thanks.

gchoqueux commented on src/Core/View.js:
```diff
@@ -0,0 +1,185 @@
+    this.onAfterRender = () => {};
+}
+
+function _preprocessLayer(view, layer, provider) {
```
`_preprocessLayer` We will see later but this function does things too specific

peppsac replied:
Agreed

gchoqueux commented on src/Core/View.js:
```diff
@@ -0,0 +1,185 @@
+    }
+}
+
+View.prototype.addLayer = function addLayer(layer, parentLayer) {
```
Note:
`addLayer` but there's not removeLayer

peppsac replied:
Yeah... mainly because the main usage is to add layers.
We'll probably want to add a `removeLayer` method layer though.

gchoqueux commented on src/Core/View.js:
```diff
@@ -0,0 +1,185 @@
+    _preprocessLayer(this, layer, this.mainLoop.scheduler.getProtocolProvider(layer.protocol));
+};
+
+View.prototype.initDefaultProviders = function initDefaultProviders(scheduler) {
```
`initDefaultProviders` move in Scheduler? or create provider when is necessary

peppsac replied:
Hmm ok, moving this to Scheduler.

gchoqueux commented on src/Core/View.js:
```diff
@@ -0,0 +1,185 @@
+    }
+};
+
+View.prototype.getLayers = function getLayers(filter) {
```
comment filter use

peppsac replied:
Done

gchoqueux commented on src/Core/Scheduler/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -420,135 +371,80 @@ ApiGlobe.prototype.createSceneGlobe = function createSceneGlobe(globeLayerId, co
+    this.layers[0].node.updateLightingPos(lightingPos);
+};
+
+ApiGlobe.prototype.animateTime = function animateTime(value) {
```
remove animateTime

peppsac replied:
Done

gchoqueux commented on src/Core/Scheduler/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -420,135 +371,80 @@ ApiGlobe.prototype.createSceneGlobe = function createSceneGlobe(globeLayerId, co
+    }
+};
+
+ApiGlobe.prototype.orbit = function orbit(value) {
```
remove orbit

peppsac replied:
Done

gchoqueux commented on src/Core/MainLoop.js:
```diff
@@ -0,0 +1,101 @@
+
+
+    if (this.scheduler.commandsWaitingExecutionCount() == 0 && executedDuringUpdate == 0) {
+        // TODO per view
```
This event is important, it can't be removed

peppsac replied:
Fixed.

peppsac commented:
New update, fixing all your comments @gchoqueux :)

vpicavet commented:
Thanks for the awesome work on this ! Now heading towards a great framework !
